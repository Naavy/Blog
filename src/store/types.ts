export interface User {
  id: string;
  first_name: string;
  last_name: string;
  gender: string;
  dob: string;
  email: string;
  phone: string;
  website: string;
  address: string;
  status: string;
  _links?: {
    self: {
      href: string;
    },
    edit: {
      href: string;
    }
    avatar: {
      href: string;
    }
  }

}

export interface Post {
  id: string;
  user_id: string;
  title: string;
  body: string;
  _links?: {
    self: {
      href: string;
    },
    edit: {
      href: string;
    }
  }
}

export interface Comment {
  id: string;
  post_id: string;
  name: string;
  email: string;
  body: string;
  _links?: {
    self: {
      href: string;
    },
    edit: {
      href: string;
    }
  }
}

export interface AppState {
  users: User[];
  posts: Post[];
  comments: Comment[];
  selectedUser?: User;
  errors?: string;
  postsLoading: boolean;
  commentsLoading: boolean;
}

