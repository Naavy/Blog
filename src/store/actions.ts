import { Dispatch } from 'redux';
import { User, Post, Comment } from './types';

export const GET_USERS = 'GET_USERS';
export const GET_USERS_SUCCESS = 'GET_USERS_SUCCESS';
export const PATCH_USER = 'PATCH_USER';
export const PATCH_USER_SUCCESS = 'PATCH_USER_SUCCESS';
export const GET_POSTS = 'GET_POSTS';
export const GET_POSTS_SUCCESS = 'GET_POSTS_SUCCESS';
export const GET_COMMENTS = 'GET_COMMENTS';
export const GET_COMMENTS_SUCCESS = 'GET_COMMENTS_SUCCESS';
export const SELECT_USER = 'SELECT_USER';
export const ADD_COMMENT = 'ADD_COMMENT';
export const ADD_COMMENT_SUCCESS = 'ADD_COMMENT_SUCCESS';
export const ADD_COMMENT_FAILURE = 'ADD_COMMENT_FAILURE';

const API = 'https://gorest.co.in/public-api';
const ACCESS_TOKEN = 'access-token=bfbcWHZ3j1wpEpMOlwuMGniVp7hQo6O3R4nt';


export const fetchUsers: any = () => {
  return (dispatch: Dispatch): void => {
    dispatch(getUsers());
    fetch(`${API}/users?${ACCESS_TOKEN}`)
      .then(response => response.json())
      .then(json => dispatch(getUsersSuccess(json.result)))
  };
};

export function getUsers() {
  return {
    type: GET_USERS
  }
}

export function getUsersSuccess(payload: User[]) {
  return {
    type: GET_USERS_SUCCESS,
    payload
  }
}

export function selectUser(payload: User) {
  return {
    type: SELECT_USER,
    payload
  }
}

export const changeUserName: any = (id: string, data: Record<string,string>) => {
  return (dispatch: Dispatch): void => {
    dispatch(patchUser(data));
    fetch(`${API}/users/${id}?${ACCESS_TOKEN}`,
      {
        method: 'PATCH',
        body: JSON.stringify(data),
        headers: {'Content-Type': 'application/json'}
      }
    )
      .then(response => response.json())
      .then(json => dispatch(patchUserSuccess(json.result)))
  };
};

export function patchUser(payload: Record<string,string>) {
  return {
    type: PATCH_USER,
    payload
  }
}

export function patchUserSuccess(payload: Record<string,string>) {
  return {
    type: PATCH_USER_SUCCESS,
    payload
  }
}

export const fetchPosts: any = (id: string) =>  {
  return (dispatch: Dispatch): void => {
    dispatch(getPosts());
    fetch(`${API}/posts?user_id=${id}&${ACCESS_TOKEN}`)
      .then(response => response.json())
      .then(json => dispatch(getPostsSuccess(json.result)))
  };
};

export function getPosts() {
  return {
    type: GET_POSTS
  }
}

export function getPostsSuccess(payload: Post[]) {
  return {
    type: GET_POSTS_SUCCESS,
    payload
  }
}

export const fetchComments: any = (id: string) =>  {
  return (dispatch: Dispatch): void => {
    dispatch(getComments());
    fetch(`${API}/comments?post_id=${id}&${ACCESS_TOKEN}`)
      .then(response => response.json())
      .then(json => dispatch(getCommentsSuccess(json.result)))
  };
};

export function getComments() {
  return {
    type: GET_COMMENTS
  }
}

export function getCommentsSuccess(payload: Comment[]) {
  return {
    type: GET_COMMENTS_SUCCESS,
    payload
  }
}

export const addComment: any = (comment: Comment) => {
  return (dispatch: Dispatch): void => {
    dispatch(addNewComment(comment));
    fetch(`${API}/comments?${ACCESS_TOKEN}`,
      {
        method: 'POST',
        body: JSON.stringify(comment),
        headers: {'Content-Type': 'application/json'}
      }
    )
      .then(response => response.json())
      .then(json => {
        if(json._meta.success) {
          dispatch(addNewCommentSuccess(json.result))
        } else { dispatch(addNewCommentFailure(json.result)) }
      })
  };
};

export function addNewComment(payload: Comment) {
  return {
    type: ADD_COMMENT,
    payload
  }
}

export function addNewCommentSuccess(payload: Comment) {
  return {
    type: ADD_COMMENT_SUCCESS,
    payload
  }
}

export function addNewCommentFailure(payload: any) {
  return {
    type: ADD_COMMENT_FAILURE,
    payload
  }
}
