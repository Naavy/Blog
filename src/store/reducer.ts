import { Reducer } from 'redux';
import { AppState } from './types';
import {
  GET_USERS_SUCCESS,
  SELECT_USER,
  GET_POSTS,
  GET_POSTS_SUCCESS,
  GET_COMMENTS,
  GET_COMMENTS_SUCCESS,
  PATCH_USER_SUCCESS,
  ADD_COMMENT_SUCCESS,
  ADD_COMMENT_FAILURE
} from './actions';
import _ from 'lodash';

const initialState: AppState = {
  users: [],
  posts: [],
  comments: [],
  selectedUser: undefined,
  errors: undefined,
  postsLoading: false,
  commentsLoading: false
}

const reducer: Reducer<AppState> = (state = initialState, action) => {
  switch (action.type) {
    case GET_USERS_SUCCESS: {
      return {
        ...state,
        users: action.payload
      };
    }
    case SELECT_USER: {
      return {
        ...state,
        selectedUser: action.payload
      }
    }
    case GET_COMMENTS: {
      return {
        ...state,
        commentsLoading: true
      };
    }

    case GET_COMMENTS_SUCCESS: {
      return {
        ...state,
        comments:  _.uniqBy([...state.comments, ...action.payload], 'id'),
        commentsLoading: false
      };
    }

    case GET_POSTS: {
      return {
        ...state,
        postsLoading: true
      };
    }

    case GET_POSTS_SUCCESS: {
      return {
        ...state,
        posts: action.payload,
        comments: [],
        postsLoading: false
      };
    }
    case PATCH_USER_SUCCESS: {
      const newUsers = state.users.map((user) => {
        if(user.id === action.payload.id) {
          user = action.payload;
        }
        return user;
      })

      return {
        ...state,
        selectedUser: action.payload,
        users: newUsers
      };
    }

    case ADD_COMMENT_SUCCESS: {
      return {
        ...state,
        comments: [...state.comments, action.payload]
      };
    }

    case ADD_COMMENT_FAILURE: {
      return {
        ...state,
        errors: action.payload.message
      }
    }

    default: {
      return state;
    }
  }
};

export { reducer };
