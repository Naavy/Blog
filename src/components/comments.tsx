import React from 'react';
import { fetchComments } from '../store/actions';
import { connect } from 'react-redux';
import { Dispatch } from 'redux';
import { AppState, Post, Comment } from '../store/types';
import { Link, Typography, Box } from '@material-ui/core';
import { Loader } from './loader';
import AddComment from './add-comment';
import _ from 'lodash';

interface State {
  showForm: boolean;
  showComments: boolean;
}

interface DispatchProps {
  getComments: (id: string) => any;
}

interface StoreProps {
  comments: Comment[];
  commentsLoading: boolean;
}

interface ComponentProps {
  post: Post
}

type Props = DispatchProps & StoreProps & ComponentProps;

export class Comments extends React.Component<Props, State> {
  readonly state: State = {
    showForm: false,
    showComments: false
  }

  getComments = () => {
    this.setState({ showComments: !this.state.showComments, showForm: false });
    this.props.getComments(this.props.post.id);
  }

  render () {
    return (
      <React.Fragment>
        {this.state.showComments &&
        <React.Fragment>
          {this.props.comments?.map((comment: Comment, index) =>
            (comment.post_id === this.props.post.id) &&
              <Box key={`${index}`} style={{ margin: '10px', backgroundColor:'#ddd', borderRadius: '10px', padding: '10px'}}>
                <Typography key={`author-${index}`} variant='body2'>
                  <b>{comment.name}</b>
                </Typography>
                <Typography key={`comment-${index}`} variant='body2'>
                  {comment.body}
                </Typography>
              </Box>
            )
          }
          {this.props.commentsLoading && <Loader />}
          {!this.props.commentsLoading &&
          _.isEmpty(this.props.comments.filter(comment => comment.post_id === this.props.post.id)) &&
            <Typography> No one has commented on this post yet </Typography>
          }
          <Link
            style={{ cursor: 'pointer', margin: '10px' }}
            onClick={() => this.setState({ showForm: !this.state.showForm}) }
          >
            <Typography variant="caption"> Add Comment </Typography>
          </Link>
        </React.Fragment>
        }
        {this.state.showForm && <AddComment post={this.props.post} />}
        <Box style={{ display: 'flex', justifyContent: 'flex-end' }}>
          <Link
            style={{ cursor: 'pointer', marginRight: '20px' }}
            onClick={this.getComments}
          >
            <Typography variant="caption"> Comments </Typography>
          </Link>
        </Box>
      </React.Fragment>
    )
  }
}

const mapStoreToProps = (state: AppState): StoreProps => {
  return {
    comments: state.comments,
    commentsLoading: state.commentsLoading
  }
}

const mapDispatchToProps = (dispatch: Dispatch): DispatchProps => {
  return {
    getComments: (id) => {
      return dispatch(fetchComments(id))
    }
  }
}

export default connect(mapStoreToProps, mapDispatchToProps)(Comments);


