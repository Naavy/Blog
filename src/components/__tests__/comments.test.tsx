import React from 'react';
import { shallow } from 'enzyme';
import { Comments } from '../comments';
import { Link, Box, Typography } from '@material-ui/core';
import { Post, Comment } from '../../store/types';

const store = {
    showForm: false,
    showComments: false,
    getComments: jest.fn,
    comments: [{}] as Comment[],
    commentsLoading: false,
    post: {} as Post
}

describe('Comments', () => {
  it('should render', () => {
    // when
    const wrapper = shallow(<Comments {...store} />);

    // then
    expect(wrapper).toMatchSnapshot();
  });

  it('should call getComments, when click on "Comments" ', () => {
    // given
    const mockedGetComments = jest.fn().mockResolvedValue(null);
    const testStore = {
      ...store,
      getComments: mockedGetComments
    }
    const wrapper = shallow(<Comments {...testStore} />);

    // when
    wrapper.find(Link).simulate('click');

    // then
    expect(mockedGetComments).toHaveBeenCalled();
  });

  it('should display only comments belonging to the post ', () => {
    // given
    const testStore = {
      ...store,
      comments: [
        {
          id: "1",
          post_id: "10",
          name: "TestName",
          email: "test@mail.com",
          body: "Test"
        },
        {
          id: "2",
          post_id: "10",
          name: "TestName2",
          email: "test2@mail.com",
          body: "Test2"
        },
        {
          id: "3",
          post_id: "9",
          name: "TestName3",
          email: "test3@mail.com",
          body: "Test3"
        }
      ],
      post: {
        id: "10",
        user_id: "1",
        title: "Test-Title10",
        body: "Test10"
      }
    }

    // when
    const wrapper = shallow(<Comments {...testStore} />);
    wrapper.setState({ showComments: true });

    // then
    expect((wrapper).find(Box)).toHaveLength(3);
  });

  it('should display text, when post does not have any comments', () => {
    // given
    const testStore = {
      ...store,
      comments: [
        {
          id: "1",
          post_id: "7",
          name: "TestName",
          email: "test@mail.com",
          body: "Test"
        },
        {
          id: "2",
          post_id: "8",
          name: "TestName2",
          email: "test2@mail.com",
          body: "Test2"
        },
        {
          id: "3",
          post_id: "9",
          name: "TestName3",
          email: "test3@mail.com",
          body: "Test3"
        }
      ],
      post: {
        id: "10",
        user_id: "1",
        title: "Test-Title10",
        body: "Test10"
      }
    }

    // when
    const wrapper = shallow(<Comments {...testStore} />);
    wrapper.setState({ showComments: true });

    // then
    expect((wrapper).find(Box)).toHaveLength(1);
    expect((wrapper).find(Typography).at(0).text()).toEqual(" This post has not yet commented ");
  });
});