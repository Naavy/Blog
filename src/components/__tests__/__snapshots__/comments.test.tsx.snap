// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comments should render 1`] = `ShallowWrapper {}`;
